shpc/main/registry/github.py

```diff
@@ -0,0 +1,174 @@
+__author__ = "Vanessa Sochat"
+__copyright__ = "Copyright 2021-2022, Vanessa Sochat"
+__license__ = "MPL 2.0"
+
+
+import shpc.utils
+from shpc.logger import logger
+import requests
+import subprocess as sp
+import re
+import os
+import sys
+
+from .provider import Provider, Result
+
+
+class RemoteResult(Result):
+    """
+    A remote result provides courtesy functions for interacting with
+    a container yaml recipe on GitHub.
+    """
+
+    def __init__(self, module, spec, load=True, config=None):
+        self.module = module
+        self._spec = spec
+        self._config = config
+        if load:
+            self.load(spec)
+
+    def load(self, spec):
+        """
+        Load the settings file into the settings object
+        """
+        if "config" not in spec:
+            logger.exit("Cannot find required 'config' key in remote metadata.")
+        self.package_file = spec["config_url"]
+        self._config = spec["config"]
+
+    def get_overrides(self, tag):
+        """
+        Load a filesystem based override file
+        """
+        logger.exit(
+            "We don't have support for remote overrides yet! Please open an issue!"
+        )
+
+    def save(self, package_file):
+        raise ValueError("Remote save to a GitHub registry is not supported.")
+
+    def load_wrapper_script(self, container_tech, script):
+        """
+        Get the content of a wrapper script, if it exists.
+        """
+        wrapper_script = self.find_wrapper_script(container_tech, script)
+
+        if wrapper_script:
+            url = os.path.join(self.dirname, wrapper_script)
+            response = requests.get(url)
+            if response.status_code != 200:
+                logger.warning(
+                    "Could not find wrapper script %s in remote registry." % script
+                )
+            return response.text
+
+    def override_exists(self, tag):
+        """
+        Check if an override exists.
+        """
+        logger.exit(
+            "We don't have support for remote overrides yet! Please open an issue!"
+        )
+
+
+class GitHub(Provider):
+    """
+    Currently a GitHub provider is only needed for on-demand upgrade.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.tag = kwargs.get("tag")
+
+        # Cache of remote container metadata
+        self._cache = {}
+
+        # E.g., subdirectory with registry files
+        self.subdir = kwargs.get("subdir")
+        super().__init__(*args, **kwargs)
+        self._url = self.source
+
+    def exists(self, name):
+        """
+        Determine if a module exists in the registry.
+        """
+        dirname = self.source
+        if self.subdir:
+            dirname = os.path.join(dirname, self.subdir)
+        return os.path.exists(os.path.join(dirname, name))
+
+    @classmethod
+    def matches(cls, source):
+        return cls.__name__.lower() in source and source.startswith("http")
```

Comment:
This made shpc reject my gitlab-based repository of container.yaml. Do you need specific features from GitHub ?

Reply:
Nope, I think we could support both, it's just a clone.

Reply:
Ah - so I take it back - there is a GitHub pages component. If we can get a working version to serve the same kind of Github pages (does GitLab have that?) I can develop this backend. See under main/registry/github.py the function called _update_cache.